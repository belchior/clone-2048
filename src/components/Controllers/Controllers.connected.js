import { connect } from 'react-redux';

import { Controllers } from './Controllers';
import { ROLLBACK, SAVE, LOAD, RESTART, TOGGLE_HARDMODE } from '../../actions/types';

export const mapStateToProps = (state) => {
  return {
    rollBack: state.rollBack,
    hardMode: state.hardMode,
  };
};

export const mapDispatchToProps = (dispatch) => {
  return {
    rollBackAction: () => dispatch({type: ROLLBACK}),
    saveAction: () => dispatch({type: SAVE}),
    loadAction: () => dispatch({type: LOAD}),
    restartAction: () => dispatch({type: RESTART}),
    hardModeAction: () => dispatch({type: TOGGLE_HARDMODE}),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Controllers);
