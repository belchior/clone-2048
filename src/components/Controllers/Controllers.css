@import '../../static/vars.css';

@media all {
  .Controllers {
    display: flex;
    flex-wrap: wrap;
    justify-content: space-between;
    padding: 2rem 1rem 4rem 1rem;
  }
  .Controllers .subtitle {
    flex: 100%;
  }
  .Controllers .btn {
    flex: 1 0 7rem;
    margin: 0 1rem 1rem 0;
  }
  .Controllers .btn:nth-child(2n+1) {
    margin-right: 0;
  }
  .Controllers .btn:nth-child(n+4) {
    margin-bottom: 0;
  }
  .Controllers .btn-toggle-label {
    flex: 100%;
    margin: 1rem 0 0 0;
  }
}
