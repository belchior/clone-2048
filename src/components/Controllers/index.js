export * from './Controllers';
export * from './Controllers.connected';
export { default } from './Controllers.connected';
