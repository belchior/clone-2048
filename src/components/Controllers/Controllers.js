import React from 'react';
import PropTypes from 'prop-types';

import { WELCOME } from '../../reducers/actions/types';
import { ButtonToggle } from '../ButtonToggle';
import './Controllers.css';

export function Controllers(props) {
  const {
    hardMode,
    hardModeAction,
    restartAction,
    rollback,
    rollbackAction,
    status,
  } = props;

  return (
    <div className="Controllers">
      <h2 className="subtitle">Controls</h2>
      <button
        className="btn controls-btn_back"
        disabled={status === WELCOME || rollback === 0}
        onClick={rollbackAction}
        title="Step back"
        type="button"
      >
        back ({rollback})
      </button>
      <button
        className="btn controls-btn_restart"
        disabled={status === WELCOME}
        onClick={restartAction}
        title="Restart the game"
        type="button"
      >
        restart
      </button>
      <ButtonToggle
        active={hardMode}
        label="Hard mode"
        onClick={hardModeAction}
        title="Hard mode"
      />
    </div>
  );
}

Controllers.propTypes = {
  hardMode: PropTypes.bool,
  hardModeAction: PropTypes.func.isRequired,
  restartAction: PropTypes.func.isRequired,
  rollback: PropTypes.number,
  rollbackAction: PropTypes.func.isRequired,
  status: PropTypes.string.isRequired,
};

Controllers.defaultProps = {
  hardMode: false,
  rollback: 0,
};
