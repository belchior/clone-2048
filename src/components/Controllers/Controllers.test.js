import React from 'react';
import { Controllers } from './Controllers';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

const rollbackAction = jest.fn();
const restartAction = jest.fn();
const hardModeAction = jest.fn();

it('should render Controllers without crashing', () => {
  let tree = renderer.create(
    <Controllers
      rollbackAction={rollbackAction}
      restartAction={restartAction}
      hardModeAction={hardModeAction}
      hardMode={false}
      rollback={2}
    />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should add disabled attribute when rollback is 0', () => {
  let tree = renderer.create(
    <Controllers
      rollbackAction={rollbackAction}
      restartAction={restartAction}
      hardModeAction={hardModeAction}
      hardMode={false}
      rollback={0}
    />
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

describe('when clicked on action buttons', () => {
  let tree;

  beforeEach(() => {
    tree = shallow(
      <Controllers
        rollbackAction={rollbackAction}
        restartAction={restartAction}
        hardModeAction={hardModeAction}
        hardMode={false}
        rollback={2}
      />
    );
  });

  it('should rollbackAction be called', () => {
    tree.find('.controls-btn_back').simulate('click');
    expect(rollbackAction.mock.calls).toHaveLength(1);
  });

  it('should restartAction be called', () => {
    tree.find('.controls-btn_restart').simulate('click');
    expect(restartAction.mock.calls).toHaveLength(1);
  });
});
