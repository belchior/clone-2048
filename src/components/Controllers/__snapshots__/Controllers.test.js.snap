// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Controllers with buttons disabled without crashing 1`] = `
<div
  className="Controllers"
>
  <h2
    className="subtitle"
  >
    Controls
  </h2>
  <button
    className="btn controls-btn_back"
    disabled={true}
    onClick={[Function]}
    title="Step back"
  >
    back (
    2
    )
  </button>
  <button
    className="btn controls-btn_restart"
    disabled={true}
    onClick={[Function]}
    title="Restart the game"
  >
    restart
  </button>
  <label
    className="btn-toggle-label"
    htmlFor="toggleButton"
  >
    <button
      className="btn-toggle"
      data-active={false}
      id="toggleButton"
      onClick={[Function]}
      title="Hard mode"
    />
    Hard mode
  </label>
</div>
`;

exports[`should render Controllers with buttons enabled without crashing 1`] = `
<div
  className="Controllers"
>
  <h2
    className="subtitle"
  >
    Controls
  </h2>
  <button
    className="btn controls-btn_back"
    disabled={false}
    onClick={[Function]}
    title="Step back"
  >
    back (
    2
    )
  </button>
  <button
    className="btn controls-btn_restart"
    disabled={false}
    onClick={[Function]}
    title="Restart the game"
  >
    restart
  </button>
  <label
    className="btn-toggle-label"
    htmlFor="toggleButton"
  >
    <button
      className="btn-toggle"
      data-active={false}
      id="toggleButton"
      onClick={[Function]}
      title="Hard mode"
    />
    Hard mode
  </label>
</div>
`;

exports[`should render Controllers with rollback button disabled when rollback is 0 1`] = `
<div
  className="Controllers"
>
  <h2
    className="subtitle"
  >
    Controls
  </h2>
  <button
    className="btn controls-btn_back"
    disabled={true}
    onClick={[Function]}
    title="Step back"
  >
    back (
    0
    )
  </button>
  <button
    className="btn controls-btn_restart"
    disabled={false}
    onClick={[Function]}
    title="Restart the game"
  >
    restart
  </button>
  <label
    className="btn-toggle-label"
    htmlFor="toggleButton"
  >
    <button
      className="btn-toggle"
      data-active={false}
      id="toggleButton"
      onClick={[Function]}
      title="Hard mode"
    />
    Hard mode
  </label>
</div>
`;
